packages/python/src/components/DataclassDeclaration.tsx

```diff
@@ -0,0 +1,190 @@
+    throw new Error("Only one KW_ONLY sentinel is allowed per dataclass body");
+  }
+}
+
+/**
+ * Allowed keyword arguments for the Python `@dataclass(...)` decorator.
+ * Single source of truth: runtime keys and compile-time type are derived here.
+ */
+export const dataclassDecoratorKeys = [
+  "init",
+  "repr",
+  "eq",
+  "order",
+  "unsafeHash",
+  "frozen",
+  "matchArgs",
+  "kwOnly",
+  "slots",
+  "weakrefSlot",
+] as const;
+export const dataclassDecoratorKeySet = new Set<string>(
+  dataclassDecoratorKeys as unknown as string[],
+);
+export type DataclassDecoratorKey = (typeof dataclassDecoratorKeys)[number];
+export type DataclassDecoratorKwargs = Partial<
+  Record<DataclassDecoratorKey, boolean>
+>;
+
+export interface DataclassDeclarationProps
+  extends ClassDeclarationProps,
+    DataclassDecoratorKwargs {}
+
+/**
+ * Renders a Python dataclass.
+ *
+ * Example:
+ * ```tsx
+ * <py.DataclassDeclaration name="User" kwOnly>
+ *   <py.VariableDeclaration instanceVariable omitNone name="id" type="int" />
+ *   <py.VariableDeclaration instanceVariable name={namekey("_", { ignoreNamePolicy: true })} type={dataclassesModule["."].KW_ONLY} omitNone />
+ *   <py.VariableDeclaration
+ *     instanceVariable
+ *     name="name"
+ *     type="str"
+ *     initializer={"Anonymous"}
+ *   />
+ * </py.DataclassDeclaration>
+ * ```
+ * Will render as:
+ * ```py
+ * from dataclasses import dataclass
+ * from dataclasses import KW_ONLY
+ *
+ * @dataclass(kw_only=True)
+ * class User:
+ *     id: int
+ *     _: KW_ONLY
+ *     name: str = "Anonymous"
+ * ```
+ */
+export function DataclassDeclaration(props: DataclassDeclarationProps) {
+  // Collect flags from props in the order they appear (preserves emission order)
+  const decoratorEntries: Array<[string, any]> = [];
+  const kwargs = {} as DataclassDecoratorKwargs;
+  for (const key of Object.keys(props)) {
+    // Only include known flags; skip undefined values
+    if (dataclassDecoratorKeySet.has(key)) {
+      const value = (props as any)[key];
+      if (value !== undefined) {
+        (kwargs as any)[key] = value;
+        decoratorEntries.push([snakeCase(key), value]);
+      }
+    }
+  }
+  const hasDecoratorArgs = decoratorEntries.length > 0;
+
+  if (hasDecoratorArgs) {
+    validateDataclassDecoratorArgs(kwargs);
+  }
+
+  function RunSymbolValidation() {
+    validateDataclassMemberConflicts(kwargs as DataclassDecoratorKwargs);
+    return null;
+  }
+
+  return (
+    <>
+      {"@"}
+      {dataclassesModule["."].dataclass}
+      <Show when={hasDecoratorArgs}>
+        (
+        <For each={decoratorEntries} comma space>
+          {([k, v]) => (
+            <>
+              {k}=<Atom jsValue={v} />
+            </>
+          )}
+        </For>
+        )
+      </Show>
+      <hbr />
+      <ClassDeclaration name={props.name} bases={props.bases} doc={props.doc}>
+        <StatementList>{props.children}</StatementList>
+        <RunSymbolValidation />
```

Comment:
I think you could just do `{validateDataclassMemberConflicts(kwargs as DataclassDecoratorKwargs)}` and avoid the component.